src/main/kotlin/org/jitsi/jicofo/conference/source/ValidatingConferenceSourceMap.kt

```diff
@@ -18,67 +18,302 @@
+         * -- There are no MSID conflicts
+         */
+        @Throws(ValidationFailedException::class)
+        private fun validateEndpointSourceSet(endpointSourceSet: EndpointSourceSet) {
+            endpointSourceSet.ssrcGroups.forEach { group ->
+                if (group.ssrcs.isEmpty()) {
+                    throw IllegalStateException("Empty group should have been filtered out.")
+                }
+                var groupMsid: String? = null
+                group.ssrcs.forEach { ssrc ->
+                    // We are doing a linear search in a list, but the list only has sources for one endpoint, so
+                    // its size is limited to [maxSsrcsPerUser].
+                    val source = endpointSourceSet.sources.find { it.ssrc == ssrc }
+                        ?: throw IllegalStateException(
+                            "Groups with SSRCs that have no corresponding source should have been filtered out."
+                        )
+
+                    if (group.semantics == SsrcGroupSemantics.Fid && group.ssrcs.size != 2) {
+                        throw InvalidFidGroupException(group.ssrcs)
+                    }
+
+                    if (source.msid == null) {
+                        throw RequiredParameterMissingException("msid")
+                    }
+                    if (groupMsid == null) {
+                        groupMsid = source.msid
+                    }
+
+                    if (source.msid != groupMsid) {
+                        throw GroupMsidMismatchException(group.ssrcs)
+                    }
+                }
+            }
+
+            // This groups all sources according to the extended group that they belong to, i.e. a SIM group gets
+            // associated with all of its sources plus the sources in FID groups associated with the SIM group's
+            // sources. For example given these groups:
+            // SIM(1, 2, 3)
+            // FID(1, 4)
+            // FID(2, 5)
+            // FID(3, 6)
+            // FID(111, 222)
+            // And non-grouped source 333 this will associate the SIM group with sources 1..6, the lonely FID group
+            // with sources 111 and 222, and a dummy single-source group with source 333.
+            //
+            // This is technically O(s*g^2) where s is the number of sources and g is the number of groups. But in
+            // practice both s and g are very small (at most 20 and 4 respectively).
+            val grouped: Map<SsrcGroup, List<Source>> =
+                endpointSourceSet.sources.groupBy { groupBySimulcastGroup(it.ssrc, endpointSourceSet.ssrcGroups) }
+
+            // Here we check for MSID conflicts, i.e. we make sure that each group has a unique MSID.
+            val msidsSeen = mutableSetOf<String?>()
+            grouped.values.map { it[0] }.forEach { source ->
+                if (msidsSeen.contains(source.msid)) throw MsidConflictException(source.msid ?: "NONE")
```

Comment:
Since a source's msid is nullable, does `msidsSeen.contains(null)` make sense?

Reply:
Yes, because the set is of nullable strings. The following works as one would expect:
```
val s = mutableSetOf<String?>()

s.contains(null) //false
s.add(null)
s.contains(null) //true
```